cloudant-sync-datastore-android/src/test/java/com/cloudant/sync/replication/ReplicationServiceTest.java

```diff
@@ -459,4 +470,127 @@ else if (operationId == PeriodicReplicationService.COMMAND_STOP_REPLICATION && l
+     * {@link ReplicationService#setReplicators(Replicator[])} is called before processing
+     * the start replication operation.
+     */
+    public void testOnStartCommandStartReplicationWaitsForSetReplicators() {
+        PeriodicReplicationService service = new TestReplicationService(mMockContext);
+        Intent intent = new Intent(mMockContext, TestReplicationService.class);
+        intent.putExtra(PeriodicReplicationService.EXTRA_COMMAND, PeriodicReplicationService.COMMAND_START_REPLICATION);
+        service.setReplicationPolicyManager(mMockReplicationPolicyManager);
+
+        service.onCreate();
+        final CountDownLatch latch = new CountDownLatch(1);
+
+        service.setOperationStartedListener(new PeriodicReplicationService
+            .OperationStartedListener() {
+            @Override
+            public void operationStarted(int operationId) {
+                if (operationId == PeriodicReplicationService.COMMAND_START_REPLICATION) {
+                    latch.countDown();
+                }
+            }
+        });
+
+        service.onStartCommand(intent, 0, 0);
+        try {
+            // Wait a bit to check the OperationStartedListener isn't called.
+            latch.await(1000, TimeUnit.MILLISECONDS);
+            assertEquals("CountDownLatch should not be decremented", latch.getCount(), 1);
+            service.setReplicators(mMockReplicators);
+            latch.await();
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+    }
+
+    /**
+     * Check that when intents are sent to the {@link PeriodicReplicationService}
+     * the commands are queued until {@link ReplicationService#setReplicators(Replicator[])} is
+     * called and the queued messages are then processed.
+     */
+    public void testOnStartCommandStartReplicationCommandsQueuedBeforeSetReplicators() {
+        PeriodicReplicationService service = new TestReplicationService(mMockContext);
+        Intent startIntent = new Intent(mMockContext, TestReplicationService.class);
+        startIntent.putExtra(PeriodicReplicationService.EXTRA_COMMAND, PeriodicReplicationService.COMMAND_START_REPLICATION);
+
+        Intent stopIntent = new Intent(mMockContext, TestReplicationService.class);
+        stopIntent.putExtra(PeriodicReplicationService.EXTRA_COMMAND, PeriodicReplicationService
+            .COMMAND_STOP_REPLICATION);
+
+        service.setReplicationPolicyManager(mMockReplicationPolicyManager);
+
+        service.onCreate();
+        final CountDownLatch latch = new CountDownLatch(2);
+
+        service.setOperationStartedListener(new PeriodicReplicationService
+            .OperationStartedListener() {
+            @Override
+            public void operationStarted(int operationId) {
+                if (operationId == PeriodicReplicationService.COMMAND_START_REPLICATION) {
+                    latch.countDown();
```

Comment:
Is there any value in asserting that these commands come in the right order? Although the implementation is a `List` I guess it protects against changes to duplicate checking or something doing an insertion into the wrong place in future.
Maybe something like:
`if (latch.getCount() == 2){ latch.countDown() } else { fail("Wrong order")}` for start and similarly with a 1 count for the stop event. WDYT?